c/C90/C90.g4

```diff
@@ -0,0 +1,672 @@
+ IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+grammar C90;
+
+compilationUnit
+    :   externalDeclaration* EOF
+    ;
+
+externalDeclaration
+    :   functionDefinition
+    |   functionDefinitionKandR
+    |   functionDeclaration
+    |   'extern'? variableDeclaration
+    |   typeDeclaration ';'
+    |   typeDefinition
+    |   typeWillBeDeclared
+    ;
+
+storageFuncSpecifier
+    :   'static'
+    |   '__inline__'
+    |   'extern'
+    ;
+
+storageVarSpecifier
+    :   'register'
+    ;
+
+
+typeWillBeDeclared
+    : ('struct'|'union') Identifier ';'
+    ;
+
+functionSpecifier
+    :  '__inline'
+    ;
+
+//if typeSpecifier not spedified : default return int
+functionDeclaration
+    :   '__extension__'? gccDeclaratorExtension1* storageFuncSpecifier* functionSpecifier? gccDeclaratorExtension1* typeSpecifier? gccDeclaratorExtension1* visualExtension? Identifier function gccDeclaratorExtension2* ';'
+    ;
+
+functionDefinition
+    :   '__extension__'? gccDeclaratorExtension1* storageFuncSpecifier* functionSpecifier? gccDeclaratorExtension1* typeSpecifier? gccDeclaratorExtension1* Identifier '(' parameterList ')' compoundStatement
+    ;
+
+functionDefinitionKandR
+    :   typeSpecifier Identifier '(' varList? ')' (typeSpecifier Identifier ';')* compoundStatement
+    ;
+
+varList
+    : variableName (',' variableName)*
+    ;
+
+typeQualifier
+    :   'const'
+    |   '__const'
+    |   '__restrict'
+    |   '__restrict__'
+    |   'volatile'
+    ;
+
+typeDeclaration
+    : structDeclaration
+    | enumDeclaration
+    ;
+
+typeSpecifier
+    :   (storageVarSpecifier | typeQualifier)* typeName typeModifier* typeQualifier*
+    ;
+
+type
+    : typeName
+    | typeQualifier typeName
+    | typeDeclaration
+    ;
+
+typeName
+    : 'int'
+    | ('unsigned'|'signed')? 'long' 'int'?
+    | '__extension__'? ('unsigned'|'signed')? 'long' 'long' 'int'?
+    | 'long'? ('unsigned'|'signed') 'int'?
+    | 'short'
+    | ('unsigned'|'signed')? 'short' 'int'?
+    | 'short'? ('unsigned'|'signed') 'int'?
+    | ('unsigned'|'signed')? 'char'
+    | 'float'
+    | 'long'? 'double'
+    | 'float' '_Complex'
+    | 'long'? 'double' '_Complex'
+    | 'void'
+    | ('struct'|'union') Identifier
+    | 'enum' Identifier
+    | '__builtin_va_list'
+    | Identifier
+    ;
+
+typeModifier
+    : '*'
+    ;
+
+fixedParameterList
+    : parameter (',' parameter)*
+    ;
+
+parameterList
+    : fixedParameterList
+    | fixedParameterList ',' '...'
+    | 'void'
+    | /*empty*/
```

Comment:
Please don't use empty alternatives (it may degrade performance and cause generator warnings). Instead, use `?` operator in use sites.